### ISSUE TITLE
Generative Trees

### ISSUE THREAD

**Issue by 0xNeat** [neatwork-ai/neatcoder#3]
The core idea behind generative trees is to break down a task into smaller chunks and and distribute the each task to semi-independent AI workers.

When given a broad objective, GPTs will typically break down the task into a set of items and develop on each item. Given the max token constraints, LLMs will typically condition their responses to a limited set of tokens per item. In other words, GPT distributes the available token resources per item seemingly in a breath-first approach.

This means that, in order to LLMs to tackle a broad object they have to be iteratively prompted in order to increase the depth of work for each item.

## Break-down vs. Solve

Applying the same strategy recursively across the tree stack is not the optimal solution given that it will cause the work tree to be wide and shallow. 

Each worker has to make a judgement on how broad the task is and decide on the strategy:
- If the problem assigned to them is too broad, output a break-down of items to work further on;
- If the problem is sufficiently narrow, output the leaf work;

#### Explicit Classification
You could ask the model directly whether a task should be subdivided or tackled directly. For instance, if your task was "Write a history of World War II," you could first ask, "Should this task be broken down further or can it be completed in a single response?"

#### Tiered Prompts
Instead of providing the complete task all at once, break it down into smaller steps. Using the history example, you could first ask for a broad overview of the World War II timeline. Then, ask for details on each major event. This way, you're helping the model understand the complexity of the task and guiding it towards a more detailed response.

#### Requesting Task Estimation
Prior to starting the actual task, ask the AI to estimate the task complexity and potential steps to solve it. For instance, "What are the potential steps to write a history of World War II and how complex would you estimate each step to be?"

or more generally:
"You are tasked with {task}. What are the potential steps to complete this task and how complex would you estimate each step to be?"

### Guided Breakdown
Directly prompt the model to break down a task into smaller steps. For instance, "Break down the task of writing a history of World War II into smaller, manageable parts."

## Global  & Shared Context
Not all tasks or sub-tasks are completely independent from each other and some of them rely on the same bits and pieces of context. As works gets performed by each node in the tree, the output gets stored in a context chain. For each new task created, a manager worker needs to keep track of the shared context and propagate the relevant context to the new workers in a consumable way (summarised in a way that fits the prompt).

## Dependency DAGs
Tasks dependent on each other, which means that not all tasks in the same node-level can be computed in parallel. Therefore  a manager worker is needed in order to keeps track of dependencies and schedules the work based on the work dependencies.
